pkg/volume/glusterfs/glusterfs.go

```diff
@@ -479,22 +510,128 @@ func (d *glusterfsVolumeDeleter) Delete() error {
+				}
+			} else {
+				glog.Errorf("glusterfs: failed to convert gid %s", gid)
+			}
+		}
+	}
+	return nil
+}
+
 func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
 	var err error
+	// Controller restart is handled here with initialized bool in plugin struct.
+	if !r.plugin.initialized {
+		err = r.ConstructBitmap()
+		if err != nil {
+			glog.Errorf("glusterfs: failed to construct allocators from existing PVs")
+
+		}
+		r.plugin.initialized = true
+	}
+
+	glog.V(1).Infof("glusterfs: allocators after restoring: %#+v ", r.plugin.gidAllocators)
+
+	var blockStart32, blockEnd32 uint32
 	if r.options.PVC.Spec.Selector != nil {
 		glog.V(4).Infof("glusterfs: not able to parse your claim Selector")
 		return nil, fmt.Errorf("glusterfs: not able to parse your claim Selector")
 	}
 	glog.V(4).Infof("glusterfs: Provison VolumeOptions %v", r.options)
-
 	cfg, err := parseClassParameters(r.options.Parameters, r.plugin.host.GetKubeClient())
 	if err != nil {
 		return nil, err
 	}
 	r.provisioningConfig = *cfg
-
 	glog.V(4).Infof("glusterfs: creating volume with configuration %+v", r.provisioningConfig)
-	glusterfs, sizeGB, err := r.CreateVolume()
+	scName := r.options.PVC.Annotations[annStorageClass]
+	// TODO :Release GIDs
+	if startGid32, err := strconv.ParseUint(r.provisioningConfig.gidMin, 10, 32); err == nil {
+		blockStart32 = uint32(startGid32)
+	} else {
+		glog.Errorf("glusterfs: failed to parse gidMin from storage class %s parameter list", scName)
+		//TODO: Do we need to return ?
+	}
+	if endGid32, err := strconv.ParseUint(r.provisioningConfig.gidMax, 10, 32); err == nil {
+		blockEnd32 = uint32(endGid32)
+	} else {
+		glog.Errorf("glusterfs: failed to parse gidMin from storage class %s parameter list", scName)
+		//TODO: Do we need to return ?
+	}
+	glog.V(2).Infof("glusterfs: provisioning config and gid range %d [%d - %d]", r.provisioningConfig, blockStart32, blockEnd32)
+	if scName != "" {
+		if r.plugin.gidAllocators[scName] != nil {
+			glog.V(1).Infof("glusterfs: bitmap : %v exist for %s ", r.plugin.gidAllocators[scName], scName)
+		} else {
+			r.plugin.gidAllocators[scName], err = gidPoolCreate(blockStart32, blockEnd32)
+			if err != nil {
+				glog.Errorf("glusterfs: gid pool creation failed %v", err)
+				return nil, fmt.Errorf("glusterfs: gid pool creation failed: %v.", err)
+			}
+			glog.V(1).Infof("glusterfs: bitmap : %v created  for %s ", r.plugin.gidAllocators[scName], scName)
+		}
+	}
+	glog.V(1).Infof("glusterfs: current gid allocator details %#+v", r.plugin.gidAllocators[scName])
+	gidblock, err := r.plugin.gidAllocators[scName].AllocateNext()
+	if err != nil {
+		glog.Errorf("glusterfs: gid allocation error %v", err)
+		return nil, fmt.Errorf("glusterfs: gid allocation error: %v.", err)
+	}
+	reqGid := int64(gidblock.Start)
+	glog.V(2).Infof("glusterfs: assigned gid [%d] to PVC %s", reqGid, r.options.PVC.Name)
```

Comment:
We need to be able to distinguish between out-of-range errors on Release() and "I actually encountered an error releasing". Since min/max can change and put previously allocated gids outside the controlled range, if we try to release those gids later, we need to be able to ignore an out-of-range error and continue deleting the PV